### ISSUE TITLE
LANGCHAIN

### ISSUE THREAD

**Issue by omsoni1357-beep** [omsoni1357-beep/SEM-V-MAJOR-V#1]
# color_translator.py
import re
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
import torch
from datetime import datetime
import os

# Models
EN_HI_MODEL = "Helsinki-NLP/opus-mt-en-hi"
HI_EN_MODEL = "Helsinki-NLP/opus-mt-hi-en"

# ANSI color codes
BLUE = '\033[94m'
GREEN = '\033[92m'
RESET = '\033[0m'

HISTORY_FILE = "translation_history.txt"

# Split text into sentences
def split_text_into_sentences(text):
    return [s for s in re.split(r'(?<=[.!?])\s+', text.strip()) if s]

# Chunk sentences
def create_chunks(sentences, max_len=300):
    chunks, current = [], ""
    for s in sentences:
        if len(current) + len(s) + 1 <= max_len:
            current = (current + " " + s).strip() if current else s
        else:
            chunks.append(current)
            current = s
    if current:
        chunks.append(current)
    return chunks

# Translate chunked text
def translate_chunks(text, model, tokenizer, device):
    sentences = split_text_into_sentences(text)
    chunks = create_chunks(sentences)
    results = []
    for c in chunks:
        inputs = tokenizer([c], return_tensors="pt", padding=True).to(device)
        with torch.no_grad():
            out = model.generate(**inputs, max_length=450, num_beams=6, early_stopping=True)
        results.append(tokenizer.batch_decode(out, skip_special_tokens=True)[0])
    return " ".join(results)

# Load model
def load_model_for_direction(direction, device):
    if direction == "en-hi":
        tokenizer = AutoTokenizer.from_pretrained(EN_HI_MODEL)
        model = AutoModelForSeq2SeqLM.from_pretrained(EN_HI_MODEL).to(device)
    else:
        tokenizer = AutoTokenizer.from_pretrained(HI_EN_MODEL)
        model = AutoModelForSeq2SeqLM.from_pretrained(HI_EN_MODEL).to(device)
    return model, tokenizer

# Detect language
def detect_language(text):
    hindi_count = sum(0x0900 <= ord(c) <= 0x097F for c in text)
    return "hi" if hindi_count > len(text)//2 else "en"

# Save translation to history file
def save_to_file(src, trg):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    with open(HISTORY_FILE, "a", encoding="utf-8") as f:
        f.write(f"[{timestamp}] {src} → {trg}\n")

# Clear history file
def clear_history():
    if os.path.exists(HISTORY_FILE):
        os.remove(HISTORY_FILE)
        print("🗑️ History cleared!\n")

# Main program
def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    
    banner = """
    ***************************************
        🌈 Color Translator 🌈
      English ↔️ Hindi Console App
    ***************************************
    """
    print(banner)
    print(f"Device: {device}")
    print("Type 'exit' to quit, 'clear' to clear history.\n")

    while True:
        text = input("Enter text: ").strip()
        if text.lower() in ("exit", "quit"):
            print("👋 Exiting translator. Goodbye!")
            break
        if text.lower() == "clear":
            clear_history()
            continue
        if not text:
            print("⚠️ Please enter some text.")
            continue

        # Detect language
        lang = detect_language(text)
        direction = "en-hi" if lang == "en" else "hi-en"
        model, tokenizer = load_model_for_direction(direction, device)

        try:
            translation = translate_chunks(text, model, tokenizer, device)
            timestamp = datetime.now().strftime("%H:%M:%S")
            
            # Colored output
            if direction == "en-hi":
                print(f"⏰ [{timestamp}] {BLUE}English:{RESET} {text}")
                print(f"➡️ {GREEN}Hindi:{RESET} {translation}\n")
            else:
                print(f"⏰ [{timestamp}] {GREEN}Hindi:{RESET} {text}")
                print(f"➡️ {BLUE}English:{RESET} {translation}\n")
            
            # Save translation
            save_to_file(text, translation)

        except RuntimeError as e:
            print("⚠️ Runtime error (GPU OOM?). Try shorter text or CPU.")
        except Exception as e:
            print("❌ Translation failed:", str(e))

if __name__ == "__main__":
    main()
